### PR TITLE
Nested server structure member shapes targeting simple shapes with default trait

### PR DESCRIPTION
**Pull request by 82marbag**
Closes: #2343 

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: CHANGELOG.next.toml [hunk 1]
```diff
@@ -10,3 +10,9 @@
 # references = ["smithy-rs#920"]
 # meta = { "breaking" = false, "tada" = false, "bug" = false, "target" = "client | server | all"}
 # author = "rcoh"
+
+[[smithy-rs]]
+message = "Mark unconstrained structures with optional _or_ default members."
```
Comment (david-perez) [inline CHANGELOG.next.toml:15]: This message is not useful to users at all. It makes no mention of the bug itself.

How about something like "Fix bug whereby nested server structure member shapes targeting simple shapes with `@default` produced Rust code that did not compile"?

File: CHANGELOG.next.toml [hunk 2]
```diff
@@ -10,3 +10,9 @@
...
+references = ["smithy-rs#2352"]
```
Comment (david-perez) [inline CHANGELOG.next.toml:16]: Reference the issue too: https://github.com/awslabs/smithy-rs/issues/2343.

**Review by david-perez — CHANGES_REQUESTED** [review]
Fix looks good, but we need tests so that we don't regress on this.

We need _at least_ a regular integration test / Kotlin unit test using `serverIntegrationTest` testing a minimal version of the model in https://github.com/awslabs/smithy-rs/issues/2343.

We should add a test to `ConstraintsTest` to test that indeed `@default` is now not detected as constrained on a simple shape too.

**Comment by 82marbag** [comment]
Tests have been added

File: codegen-server/src/test/kotlin/software/amazon/smithy/rust/codegen/server/smithy/ConstraintsTest.kt [hunk 3]
```diff
@@ -119,4 +125,9 @@ class ConstraintsTest {
         mapB.canReachConstrainedShape(model, symbolProvider) shouldBe true
         recursiveShape.canReachConstrainedShape(model, symbolProvider) shouldBe true
     }
+
+    @Test
+    fun `it should not consider shapes with the default trait as constrained`() {
+        structWithInnerDefault.canReachConstrainedShape(model, symbolProvider) shouldBe false
```
Comment (david-perez) [inline codegen-server/src/test/kotlin/software/amazon/smithy/rust/codegen/server/smithy/ConstraintsTest.kt:131]: I would call `isDirectlyConstrained` on `StructWithInnerDefault$PrimitiveBoolean` too, to make the unit tests be more unit-like.

File: codegen-server/src/test/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerBuilderConstraintViolationsTest.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+import org.junit.jupiter.api.Test
+import software.amazon.smithy.rust.codegen.core.testutil.asSmithyModel
+import software.amazon.smithy.rust.codegen.server.smithy.testutil.serverIntegrationTest
+
+class ServerBuilderConstraintViolationsTest {
+
+    @Test
+    fun `it should not generate constraint violations for members with a default value`() {
+        val model = """
```
Comment (david-perez) [inline codegen-server/src/test/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerBuilderConstraintViolationsTest.kt:16]: Nice. I'd leave a comment linking to https://github.com/awslabs/smithy-rs/issues/2343 so the reader can easily find how this test came to be.

In general I like how the Rust compiler adds test for reported issues and links back to the info on GitHub e.g. https://github.com/rust-lang/rust/tree/master/tests/ui/codegen

**Review by david-perez — APPROVED** [review]

**Pull request merged**